### PR TITLE
Fix: casing of unicode characters

### PR DESCRIPTION
**Pull request by armano2**
#688

### CODE REVIEW

File: tests/lib/utils/casing.js [hunk 1]
```diff
@@ -30,30 +34,56 @@ describe('getConverter()', () => {
...
+    assert.equal(converter('Foo1BAZ'), 'foo1_b_a_z')
     assert.equal(converter('foo1b_a_z'), 'foo1b_a_z')
+    assert.equal(converter('darИībaÊÊw'), 'dar_иība_ê_êw')
+    assert.equal(converter('    foo       Bar   '), 'foo_bar')
+  })
+
+  it('unicode tests', () => {
+    const words = [
+      ['ÊtreSîne', 'être_sîne'],
+      ['darbībaÊÊw', 'darbība_ê_êw'],
```
Comment (ota-meshi) [inline tests/lib/utils/casing.js]: This is correct with a popular casing. However, I think there are times when there is a difference with the solution method of Vue.
https://github.com/vuejs/vue/blob/master/src/shared/util.js#L149

I'm worried about being broken by some autofix.

Comment (armano2) [reply tests/lib/utils/casing.js]: hmm, how it should work than xd

Comment (armano2) [reply tests/lib/utils/casing.js]: vue is not checking for utf characters, they will have to polyfill.

Comment (armano2) [reply tests/lib/utils/casing.js]: @ota-meshi maybe we should just ignore non a-z utf characters?

Comment (ota-meshi) [reply tests/lib/utils/casing.js]: It's difficult to decide.
@michalsnik what do you think?

Comment (ota-meshi) [reply tests/lib/utils/casing.js]: In most of Japan, it is forbidden to use a source other than a-z to avoid problems.

Comment (ota-meshi) [reply tests/lib/utils/casing.js]: @armano2 
> what do you think about making new rule which forbids usage of non ascii characters in code?

I think that rule is very nice!

Comment (mysticatea) [reply tests/lib/utils/casing.js]: I think that we should follow the original compiler's behavior.

> what do you think about making new rule which forbids usage of non ascii characters in code?

Sounds good to me.

Comment (michalsnik) [reply tests/lib/utils/casing.js]: Perhaps if we introduce this rule and put it in the `base` category - we wouldn't need this fix at all?

**Comment by armano2** [comment]
What i should do with this than? Remove support for utf-8 and add rule to ban non a-zA-Z characters instead?

i feel like its base on preference, some ppl may want to use non ascii characters like #688

**Comment by mysticatea** [comment]
I think that:

- the utilities have to behave as same as Vue.js's. 
- the rule which disallows non-ascii *upper case* characters in names would be nice because it can cause unexpected behavior.

**Comment by michalsnik** [comment]
@mysticatea or @ota-meshi can you also take a look at the code here? I would appreciate second pair of eyes before merging it :)

**Comment by mysticatea** [comment]
My opinion is not changed since https://github.com/vuejs/eslint-plugin-vue/pull/694#issuecomment-443990334. I meant that our utilities should behave as same as https://github.com/vuejs/vue/blob/dev/src/shared/util.js. That doesn't have special handling for non-ASCII characters.

**Comment by armano2** [comment]
@mysticatea what should happen with non `A-Za-z:` characters than?

**Comment by mysticatea** [comment]
@armano2 Why does it remove those?

**Comment by armano2** [comment]
no clue, it was copied from vue :)

**Comment by armano2** [comment]
i will replace it with `.replace(/[\-!#$%^&*()_+~`\s\r\n\v\t]+/gu, ' ')`

**Comment by armano2** [comment]
and now it's ignoring unknown (non ascii) characters, and we will have to add new rule to warn about usage of non ascii characters.

**Comment by mysticatea** [comment]
> no clue, it was copied from vue :)

In that case, I think we should keep it as is. My concern is that static analysis gets different behavior to Vue.js compiler and overlooks errors.

**Comment by armano2** [comment]
@mysticatea but in vue its used in different context https://github.com/vuejs/vue/blob/a7658e03a16dc507f0abeba41aee705f773727d0/src/core/util/options.js#L260
and from what i see it got removed anyway

**Comment by mysticatea** [comment]
@mysticatea My question in https://github.com/vuejs/eslint-plugin-vue/pull/694#issuecomment-450610078 was "why remove?". But https://github.com/vuejs/vue/blob/a7658e03a16dc507f0abeba41aee705f773727d0/src/core/util/options.js#L260 is not removal. So I should ask again, why does it remove those? If the removal is different behavior to Vue.js compiler, we should not remove those characters.

**Comment by armano2** [comment]
that's why i changed it now, we are removing now only invalid characters 
```ts
\-!#$%^&*()_+~`"'<>,./?\[\]{}\s\r\n\v\t
```
those characters can't be used as attribute names and property names

**Comment by mysticatea** [comment]
Why does remove invalid characters? I think that we should handle invalid names as invalid names because this is an error checker.

**Comment by armano2** [comment]
we have to somehow determine when word ends, and those are characters on which we split text, vue compile is never removing invalid characters, it's going to be runtime error..

**Pull request closed without merging**